pkg/resourcemanager/mysql/aadadmin/reconcile.go

```diff
@@ -0,0 +1,218 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package aadadmin
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/go-autorest/autorest/to"
+	uuid "github.com/satori/go.uuid"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/Azure/azure-sdk-for-go/services/mysql/mgmt/2017-12-01/mysql"
+
+	"github.com/Azure/azure-service-operator/api/v1alpha1"
+	azurev1alpha1 "github.com/Azure/azure-service-operator/api/v1alpha1"
+	"github.com/Azure/azure-service-operator/pkg/errhelp"
+	"github.com/Azure/azure-service-operator/pkg/helpers"
+	"github.com/Azure/azure-service-operator/pkg/resourcemanager"
+	"github.com/Azure/azure-service-operator/pkg/resourcemanager/config"
+	"github.com/Azure/azure-service-operator/pkg/resourcemanager/iam"
+)
+
+type MySQLServerAdministratorManager struct {
+	Creds config.Credentials
+}
+
+func NewMySQLServerAdministratorManager(creds config.Credentials) *MySQLServerAdministratorManager {
+	return &MySQLServerAdministratorManager{
+		Creds: creds,
+	}
+}
+
+func newMySQLServerAdministratorsClient(creds config.Credentials) mysql.ServerAdministratorsClient {
+	adminClient := mysql.NewServerAdministratorsClientWithBaseURI(config.BaseURI(), creds.SubscriptionID())
+	a, _ := iam.GetResourceManagementAuthorizer(creds)
+	adminClient.Authorizer = a
+	// TODO: This is safe to ignore but still feels icky
+	_ = adminClient.AddToUserAgent(config.UserAgent())
+	return adminClient
+}
+
+func (m *MySQLServerAdministratorManager) Ensure(ctx context.Context, obj runtime.Object, opts ...resourcemanager.ConfigOption) (bool, error) {
+	instance, err := m.convert(obj)
+	if err != nil {
+		return true, err
+	}
+
+	instance.Status.SetProvisioning("")
+	client := newMySQLServerAdministratorsClient(m.Creds)
+
+	admin, err := client.Get(ctx, instance.Spec.ResourceGroup, instance.Spec.Server)
+	if err == nil {
+		instance.Status.SetProvisioned(resourcemanager.SuccessMsg)
+		instance.Status.State = admin.Status
+		return true, nil
+	}
+
+	tenantUUID, err := uuid.FromString(instance.Spec.TenantId)
+	if err != nil {
+		instance.Status.SetFailedProvisioning(fmt.Sprintf("failed to parse tenant UUID %s", instance.Spec.TenantId))
+		// TODO: Do we have to manage state here?
+		// instance.Status.State = admin.Status
```

Comment:
It seems like we'd always want the status to be the most up-to-date view we have, right? It's not bad the way it is, but if there were more early returns and you wanted to avoid the repetition you could use a defer.

Reply:
The problem I'm having is it's actually unclear to me what the `State` field even is... and I see most resources don't consistently use it.

Reply:
I've fixed this to manually setting Succeeded/Failed which is what a bunch of resources seem to do...